File: hybrid/lib/@core/models/contest_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';

class ContestModel {
  final String id;
  final String name;
  final int amount;
  final int payout;
  final String payoutLabel;
  final Timestamp startDate;
  final Timestamp endDate;
  final String courseId;
  final List<ContestModel> contests;

  ContestModel(
      {@required this.id,
      @required this.name,
      @required this.amount,
      @required this.payout,
      @required this.payoutLabel,
      @required this.startDate,
      @required this.endDate,
      @required this.courseId,
      @required this.contests});

  ContestModel.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        name = json['name'],
        amount = json['amount'],
        payout = json['payout'],
        payoutLabel = json['payoutLabel'],
        startDate = _timestampFromJson(json['startDate']),
        endDate = _timestampFromJson(json['endDate']),
        courseId = json['courseId'],
        contests = _buildContests(json['contests']);

  static Timestamp _timestampFromJson(dynamic json) {
    if (json == null) {
      return null;
    }
    return Timestamp(json['_seconds'], json['_nanoseconds']);
  }

  static List<ContestModel> _buildContests(List<dynamic> json) {
    if (json == null || json.length <= 0) {
      return null;
    }
    var ret = List<ContestModel>();
    for (var i = 0; i < json.length; i++) {
      ret.add(ContestModel.fromJson(json[i]));
    }
    return ret;
  }
}

File: iOS/app/src/Core/Models/ContestModel.swift
//
//  ContestModel.swift
//  BigMoneyShot
//
//  Created by Brian Ogden on 2/24/19.
//  Copyright © 2019 Big Money Shot. All rights reserved.
//

import Firebase

struct ContestModel: Codable {
    var id: String?
    var name: String
    var amount: Int?
    var payout: Int
    var payoutLabel: String?
    var startDate: Timestamp?
    var endDate: Timestamp? //prepped for removal in v.1.5, v1.4 has dependency
    var courseId: String?
    var courseModel: CourseModel?
    var contests: [ContestModel]?
    
    private enum CodingKeys: String, CodingKey {
        case name
        case amount
        case payout
        case payoutLabel
        case startDate
        case courseId
        case contests
    }
}
